superset/db_engine_specs/ocient.py

```diff
@@ -215,32 +227,62 @@ class OcientEngineSpec(BaseEngineSpec):
+                
+        if len(rows) > 0 and type(rows[0]).__name__ == rows:
+            # Peek at the schema to determine which column values, if any,
+            # require sanitization.
+            columns_to_sanitize: List[PlacedSanitizeFunc] = _find_columns_to_sanitize(cursor)
 
-        if (len(rows)) == 0:
-            # No rows were produced
-            return rows
+            if columns_to_sanitize:
+                # At least 1 column has to be sanitized.
+                for row in rows:
+                    for info in columns_to_sanitize:
+                        # Modify the element in-place.
+                        v = row[info.column_index]
+                        row[info.column_index] = info.sanitize_func(v)
 
-        if type(rows[0]).__name__ != 'Row':
-            # TODO what else is returned here?
-            return rows
+        # # We are done with this cursor so we can safely remove it from the cache
+        OcientEngineSpec.query_id_mapping_lock.acquire()
+        del OcientEngineSpec.query_id_mapping[getattr(cursor, 'superset_query_id')]
+        OcientEngineSpec.query_id_mapping_lock.release()
+        
+        return rows
 
-        # Peek at the schema to determine which column values, if any,
-        # require sanitization.
-        columns_to_sanitize: List[PlacedSanitizeFunc] = _find_columns_to_sanitize(cursor)
 
-        if columns_to_sanitize:
-            # At least 1 column has to be sanitized.
-            for row in rows:
-                for info in columns_to_sanitize:
-                    # Modify the element in-place.
-                    v = row[info.column_index]
-                    row[info.column_index] = info.sanitize_func(v)
+    @classmethod
+    def get_cancel_query_id(cls, cursor: Any, query: Query) -> Optional[str]:
+        # Return a Non-None value
+        # If None is returned, Superset will not call cancel_query
+        return 'DUMMY_VALUE'
 
-        return rows
+
+    @classmethod
+    def handle_cursor(cls, cursor: Any, query: Query, session: Session) -> None:
+        OcientEngineSpec.query_id_mapping_lock.acquire()
+        OcientEngineSpec.query_id_mapping[query.id] = cursor.query_id    
+        OcientEngineSpec.query_id_mapping_lock.release()
+        
+        # Add the query id to the cursor
+        setattr(cursor, "superset_query_id", query.id)
+        return super().handle_cursor(cursor, query, session)
+    @classmethod
+    def cancel_query(cls, cursor: Any, query: Query, cancel_query_id: str) -> bool:
+        OcientEngineSpec.query_id_mapping_lock.acquire() 
+        if query.id in OcientEngineSpec.query_id_mapping:
+            cursor.execute(f'CANCEL {OcientEngineSpec.query_id_mapping[query.id]}')
+            # Query has been cancelled, so we can safely remove the cursor from the cache
+            del OcientEngineSpec.query_id_mapping[query.id]
+            OcientEngineSpec.query_id_mapping_lock.release()
```

Comment:
Now this pattern will certainly deadlock the application. 

Consider the case where the user types `foobar` into the editor and clicks the shiny "Submit" button? That's a silly example but any exception raised by `cursor.execute(...)` will deadlock the SQL Lab (a syntax error in that case).

Reply:
Fixed using `with`